pkg/ebpf/c/common/context.h

```diff
@@ -209,6 +208,18 @@ statfunc int init_program_data(program_data_t *p, void *ctx)
     }
 
 out:
+    if (p->event->context.policies_version != p->config->policies_version) {
```

Comment:
Actually the check is against event's policies_version and config's policies_version. `p->config` is always fetched: line [155](https://github.com/aquasecurity/tracee/pull/3906/files/f1b2d39ee0c6393818c70abf96cda272a87f30d6#diff-2d5cf85fbb13b213de3c3fc135cc46d3cc2c30d7436ba65c232991acf47d39afR155).

Reply:
But where did we set p->event.context.policies_version?

Reply:
I assume we don't need pre-assign it, since BPF_MAP_TYPE_ARRAY and BPF_MAP_TYPE_PERCPU_ARRAY are pre-allocated and zero-initialised when created.

I've made a small test to make sure:

```diff
 out:
+    if (p->event->context.policies_version == 0 || p->config->policies_version == 0)
+        bpf_printk("\nevent->policies_version is %d\n"
+                     "config->policies_version is %d\n"
+                     "for cpu id %d\n",
+                        p->event->context.policies_version,
+                        p->config->policies_version,
+                        p->event->context.processor_id);
+
     if (p->event->context.policies_version != p->config->policies_version) {
         // copy policies_config to event data
```

```
❯ sudo timeout 30 cat /sys/kernel/debug/tracing/trace_pipe > trace_output; grep -c bpf_trace_printk trace_output    
32
```

```
❯ nproc
32
```

All prints show a percpu event entry with a stale (in this case, zero-initialised) policies_version.

```
❯ head trace_output 
            code-1545838 [018] ...2. 115176.478099: bpf_trace_printk: 
event->policies_version is 0
config->policies_version is 1
for cpu id 18

            code-1566656 [026] ...21 115176.478513: bpf_trace_printk: 
event->policies_version is 0
config->policies_version is 1
for cpu id 26
```

So I believe we can rely on this array bpf map property and only fetch a new policies_config when we have a difference between event->policies_version and config->policies_version. WDYT?

Reply:
Ok, I think I see what you mean. We know that event_data_map will return a zero intialized struct on the first time we fetch it, and we know that policy version is set to 1 at start (right?), then we perform the first update.
The next events that follow are not going to touch these fields, and we can assume they are valid as long as policy version haven't changed. Is that correct?

Reply:
Ok, maybe one last concern.
Consider the following scenario: 
1. Current policies version is 2
2. Current task scope was computed with policies version 1, and recompute_scope is false
3. The previous event already updated `p->event->context.policies_version` so it is now set to 2

Now since `p->event->context.policies_version` = `p->config->policies_version` we will never set recompute_scope of the current task to 2. Isn't that right?
If so, I think we should also have policies_version as part of task_info, so we will know to which policies_version the matched_scopes value is relevant, and add the following lines after this block:

```
if (p->event->context.policies_version != p->config->policies_version) {
    // copy policies_config to event data
    long ret = bpf_probe_read_kernel(
        &p->event->policies_config, sizeof(policies_config_t), &p->config->policies_config);
    if (unlikely(ret != 0))
        return 0;

    // finally update the version and set recompute_scope to true
    p->event->context.policies_version = p->config->policies_version;
}
if (p->task_info->policies_version != p->event->context.policies_version)
    p->task_info->recompute_scope = true;
```

WDYT?

Reply:
Per event context was dealt but not for task context change. Good catch. Raising an update.